src/group_theory/double_coset.lean

```diff
@@ -0,0 +1,320 @@
+/-
+Copyright (c) 2021 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+import data.setoid.basic
+import group_theory.subgroup.basic
+import group_theory.coset
+import data.set.basic
+import tactic.group
+
+/-!
+# Double cosets
+
+This file defines double cosets for two subgroups `H K` of a group `G` and the quotient of `G` by
+the double coset relation, i.e. `H \ G / K`. We also prove that `G` can be writen as a disjoint
+union of the double cosets and that if one of `H` or `K` is the trivial group (i.e. `⊥` ) then
+this is the usual left or right quotient of a group by a subgroup.
+
+## Main definitions
+
+* `double_coset_rel`: The double coset relation defined by two subgroups `H K` of `G`.
+* `double_coset.quotient`: The quotient of `G` by the double coset relation, i.e, ``H \ G / K`.
+-/
+
+
+variables {G : Type*} [group G] {α : Type*} [has_mul α] (J: subgroup G) (g : G)
+
+namespace double_coset
+
+/--Relation defining a double coset-/
+def double_coset_rel (H K : subgroup G) : G → G → Prop :=
+λ x y, (∃ (a ∈ H) (b ∈ K), y = a * x * b)
```

Comment:
Would it be easier to define `double_coset_rel` in terms of `doset`?

For example:
```
/--The double_coset as an element of `set α` corresponding to `s a t` -/
def doset (s : set α) (a : α) (t : set α) : set α := s * {a} * t

def double_coset_rel (H K : set G) : G → G → Prop :=
λ x y, doset H x K = doset H y K

lemma rel_reflex {H K : set G} : reflexive (double_coset_rel H K) :=
λ x, rfl

lemma rel_symm {H K : set G} : symmetric (double_coset_rel H K) :=
λ x y, eq.symm

lemma rel_trans {H K : set G} : transitive (double_coset_rel H K) :=
λ x y z, eq.trans

lemma rel_is_equiv {H K : set G} : equivalence (double_coset_rel H K) :=
⟨rel_reflex, rel_symm, rel_trans⟩

/-- The setoid defined by the double_coset relation -/
def setoid (H K : set G) : setoid G :=
⟨double_coset_rel H K, rel_is_equiv⟩

/-- Quotient of `G` by the double coset relation, i.e. `H \ G / K` -/
def quotient (H K : set G) : Type* :=
quotient (setoid H K)
```